### PR TITLE
Lab-4

### PR DESCRIPTION
**Pull request by NicholeL**
Implement functions
Implement tests
Fixes #586 

Review of colleague's PR 

Changes proposed in this PR:
Implement lab 4 of go-course
Implement testing.

### CODE REVIEW

**Review by harirakr — DISMISSED** [review]
Looks good

File: 04_numeronym/nicholeL/main.go [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+func numeronyms(vals ...string) []string {
+	strArr := make([]string, len(vals))
+	for i, str := range vals {
+		strArr[i] = numberShorten(str)
+	}
+	return strArr
+}
+
+func numberShorten(word string) string {
```
Comment (juliaogris) [inline 04_numeronym/nicholeL/main.go]: I'd call this one `numeronym`

File: 04_numeronym/nicholeL/main_test.go [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+	"testing"
+)
+
+func TestMainOutput(t *testing.T) {
+	var buf bytes.Buffer
+	out = &buf
+	main()
+	expected := strconv.Quote("[a11y K8s abc]")
+	actual := strconv.Quote(buf.String())
+	if actual != expected {
```
Comment (juliaogris) [inline 04_numeronym/nicholeL/main_test.go]: Consider using testify assert (it will shorten this to one line without need for explicit failure message).

File: 04_numeronym/nicholeL/main_test.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+	}
+}
+
+func TestNumberShorten(t *testing.T) {
+	testCases := map[string]struct {
+		input    string
+		expected string
+	}{
+		"Happy case":                {"accessibility", "a11y"},
+		"length of string is three": {"asd", "asd"},
```
Comment (juliaogris) [inline 04_numeronym/nicholeL/main_test.go]: Add Unicode characters like emojis - you'll have to update your code to work with runes.

**Review by juliaogris — CHANGES_REQUESTED** [review]
Like lab 1,2 please fix up 
- commit messages with `git rebase -i HASH` And force push
- pull request description: 
   - add issue 
    - add link to review of colleagues pr
- remove unnecessary blank lines

File: 04_numeronym/nicholeL/main.go (hunk 1 above)
Comment (NicholeL) [reply 04_numeronym/nicholeL/main.go]: thanks for your advice, will update

File: 04_numeronym/nicholeL/main_test.go (hunk 2 above)
Comment (NicholeL) [reply 04_numeronym/nicholeL/main_test.go]: thanks for your advice

File: 04_numeronym/nicholeL/main_test.go (hunk 3 above)
Comment (NicholeL) [reply 04_numeronym/nicholeL/main_test.go]: thanks, will add unicode characters like emojis

File: 04_numeronym/nicholeL/main.go [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+	}
+	return strArr
+}
+
+func numeronym(word []rune) string {
+	if len(word) <= 3 {
+		return string(word)
+	}
+	count := len(word) - 2
+	return fmt.Sprintf("%c%d%c", word[0], count, word[len(word)-1])
```
Comment (anzboi) [inline 04_numeronym/nicholeL/main.go:24]: Awesome 👍

File: 04_numeronym/nicholeL/main_test.go [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+	}
+}
+
+func TestNumeronym(t *testing.T) {
+	testCases := []struct {
+		name     string
+		input    []rune
+		expected string
+	}{
+		{"Empty string", []rune{' '}, " "},
```
Comment (anzboi) [inline 04_numeronym/nicholeL/main_test.go:37]: Great test coverage 👍

File: 04_numeronym/nicholeL/main_test.go [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+		{"length of string is two", []rune{'s', 's'}, "ss"},
+		{"length of string is three", []rune{'a', 's', 'd'}, "asd"},
+		{"length of string more than 3", []rune{'d', 's', 'a', 'd', 'a', 's', 'd', 'a', 's', 'f', 'f',
+			'j', 'k', 'j', 'n', 'n', 'v', 'c', 'c', 'd', 'f'}, "d19f"},
+		{"string contain emojis and special text", []rune{'😀', '😀', '😀', '🤓', '🤓', '🤓',
+			'哈', '哈', '哈'}, "😀7哈"},
+	}
+
+	for _, testCase := range testCases {
+		expected := testCase.expected
```
Comment (anzboi) [inline 04_numeronym/nicholeL/main_test.go:48]: I really think you would benefit from running these tests as subtests.
```go
for _, testCase := range testCases {
    t.Run(testCase.name, func(t *testing.T) {
        // Test code
    })
}
```
This will run each case as a subtest and produce output of each test case as if it were in its own test.

**Review by anzboi — CHANGES_REQUESTED** [review]
Great work, almost LGTM
First, you need to link a colleagues PR that you have reviewed.

Second:
I really think you would benefit from running each test case as a subtest and get in the habit of doing so when testing many test cases, It really cleans up the test output and makes catching and debugging errors so much easier since you can easily see exactly which case failed.

**Comment by harirakr** [comment]
@NicholeL  "Fixes #586" (as mentioned in the PR) I could not find this Issue number. It appears to be the PR number. You need to create an issue by going here https://github.com/anz-bank/go-course/issues and use that Issue number in "Fixes #". When the PR is merged, the corresponding issue will be automatically closed.



> Implement functions
> Implement tests
> Fixes #586
> 
> Review of colleague's PR
> 
> Changes proposed in this PR:
> Implement lab 4 of go-course
> Implement testing.
